hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/LikeFunction.java

```diff
@@ -0,0 +1,214 @@
+
+    /** Single-symbol wildcard in SQL. */
+    private static final char ONE_SQL = '_';
+
+    /** Multi-symbol wildcard in SQL. */
+    private static final char MANY_SQL = '%';
+
+    /** Single-symbol wildcard in Java. */
+    private static final String ONE_JAVA = ".";
+
+    /** Multi-symbol wildcard in Java. */
+    private static final String MANY_JAVA = "(?s:.*)";
+
+    /** Special characters which require escaping in Java. */
+    private static final String ESCAPE_CHARACTERS_JAVA = "[]()|^-+*?{}$\\.";
+
+    private transient State state;
+
+    public LikeFunction() {
+        // No-op.
+    }
+
+    private LikeFunction(Expression<?> source, Expression<?> pattern, Expression<?> escape) {
+        super(source, pattern, escape);
+    }
+
+    public static LikeFunction create(Expression<?> source, Expression<?> pattern, Expression<?> escape) {
+        return new LikeFunction(source, pattern, escape);
+    }
+
+    @SuppressFBWarnings(value = "NP_BOOLEAN_RETURN_NULL", justification = "SQL has three-valued boolean logic")
+    @Override
+    public Boolean eval(Row row, ExpressionEvalContext context) {
+        String source = asVarchar(operand1, row, context);
+
+        if (source == null) {
+            return null;
+        }
+
+        String pattern = asVarchar(operand2, row, context);
+
+        if (pattern == null) {
+            return null;
+        }
+
+        String escape;
+
+        if (operand3 != null) {
+            escape = asVarchar(operand3, row, context);
+
+            if (escape == null) {
+                return null;
+            }
+        } else {
+            escape = null;
+        }
+
+        if (state == null) {
+            state = new State();
+        }
+
+        return state.like(source, pattern, escape);
+    }
+
+    @Override
+    public QueryDataType getType() {
+        return QueryDataType.BOOLEAN;
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_LIKE;
+    }
+
+    /**
+     * Helper class to execute LIKE function. Caches the last observed pattern to avoid constant re-compilation.
+     */
+    public static class State {
+        /** Last observed pattern. */
+        private String lastPattern;
+
+        /** Last observed escape. */
+        private String lastEscape;
+
+        /** Last Java pattern. */
+        private Pattern lastJavaPattern;
+
+        public boolean like(String source, String pattern, String escape) {
+            Pattern javaPattern = convertToJavaPattern(pattern, escape);
+
+            Matcher matcher = javaPattern.matcher(source);
+
+            return matcher.matches();
+        }
+
+        private Pattern convertToJavaPattern(String pattern, String escape) {
+            if (lastJavaPattern != null && Objects.equals(pattern, lastPattern) && Objects.equals(escape, lastEscape)) {
```

Comment:
The first condition isn't needed: `pattern` is always not null here (we can assert it). If `pattern.equals(lastPattern)` it means that we've already been here and `lastJavaPattern` is also not null.
```suggestion
            if (Objects.equals(pattern, lastPattern) && Objects.equals(escape, lastEscape)) {
```

Reply:
Removed unnecessary check.